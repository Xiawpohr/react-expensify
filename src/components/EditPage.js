import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { withTheme } from 'styled-components'
import ExpenseForm from './ExpenseForm.js'
import { editExpense, removeExpense } from '../actions/expenses.js'
import BackgroundColor from '../styles/BackgroundColor.js'
import Container from '../styles/Container.js'
import Button from '../styles/Button.js'

export const EditPage = (props) => {
  const onSubmit = (expense) => {
    props.editExpense(props.id, expense)
    props.history.push('/')
  }

  const onRemove = () => {
    props.removeExpense(props.id)
    props.history.push('/')
  }

  return (
    <div>
      <BackgroundColor color={props.theme.colors.offWhite}>
        <Container py={3} mb={3}>
          <h1>Edit Expense</h1>
        </Container>
      </BackgroundColor>
      <Container>
        <ExpenseForm
          description={props.description}
          amount={props.amount}
          createdAt={props.createdAt}
          note={props.note}
          onSubmit={onSubmit}
        />
        <Button secondary onClick={onRemove}>Remove</Button>
      </Container>
    </div>
  )
}

const mapStateToProps = (state, props) => {
  const expense = state.expenses.find(expense => expense.id === props.match.params.id)
  return { ...expense }
}

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense(id))
})

export default withTheme(withRouter(connect(mapStateToProps, mapDispatchToProps)(EditPage)))
