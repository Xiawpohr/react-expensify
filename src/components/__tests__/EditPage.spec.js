import React from 'react'
import { shallow } from 'enzyme'
import { EditPage } from '../EditPage.js'
import moment from '../__mocks__/moment.js'

let theme, wrapper, editExpense, removeExpense, history, expense, id

beforeEach(() => {
  id = 'abc'
  expense = {
    description: 'bill',
    amount: 1000,
    createdAt: moment(0).valueOf(),
    note: ''
  }
  theme = {
    colors: { offWhite: '#f7f7f7' }
  }
  editExpense = jest.fn()
  removeExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(
    <EditPage
      {...expense}
      id={id}
      theme={theme}
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
    />
  )
})

test('should render Edit Page correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense action', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense)
  expect(editExpense).toHaveBeenLastCalledWith(id, expense)
  expect(history.push).toHaveBeenLastCalledWith('/')
})

test('should handle removeExpense action', () => {
  wrapper.find('Button').simulate('click')
  expect(removeExpense).toHaveBeenCalledWith(id)
  expect(history.push).toHaveBeenLastCalledWith('/')
})
