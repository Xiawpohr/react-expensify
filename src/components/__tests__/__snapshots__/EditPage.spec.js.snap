// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit Page correctly 1`] = `
<div>
  <BackgroundColor
    color="#f7f7f7"
  >
    <Container
      mb={3}
      py={3}
    >
      <h1>
        Edit Expense
      </h1>
    </Container>
  </BackgroundColor>
  <Container>
    <ExpenseForm
      amount={1000}
      createdAt={0}
      description="bill"
      note=""
      onSubmit={[Function]}
    />
    <Button
      onClick={[Function]}
      secondary={true}
    >
      Remove
    </Button>
  </Container>
</div>
`;
